---
title: "1. Preprocess"
author: "Anders Kiledal"
date: "6/9/2020"
output: html_document
---

Adapted from https://benjjneb.github.io/dada2/ITS_workflow.html

## Setup

```{r setup}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(dada2,warn.conflicts = FALSE)
library(ShortRead)
library(Biostrings)
library(foreach)
#library(doMC)
library(doSNOW)
library(here)
library(tidyverse)

knitr::opts_knit$set(root.dir = here())
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))
```

## Remove primers and ambiguous bases

List the files
```{r list_files}
path <- file.path(getwd(), "data/raw")
list.files(path)
```
Sort forward and reverse reads
```{r}
fnFs <- sort(list.files(path, pattern = "_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2_001.fastq", full.names = TRUE))
```

Define the primer sequences used
```{r define_primers}
FWD <- "CCTACGGGNGGCWGCAG"
REV <- "GACTACHVGGGTATCTAATCC"
```

Process primer seqs and get all orientations
```{r all_primer_orients}
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
        RevComp = reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
```


Filter sequences with ambiguous bases
```{r filter_ambiguous, cache=TRUE}
fnFs.filtN <- file.path(path, "filtN", basename(fnFs)) # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(path, "filtN", basename(fnRs))
filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE)

#use if want to skip filtering ambiguous bases
#fnFs.filtN <- fnFs
#fnRs.filtN <- fnRs
```


Count the number of times the primers appear in forward and reverse reads
```{r, cache=TRUE}
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.filtN[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.filtN[[1]]))
```


Primers are trimmed with cutadapt. Cutadapt is a dependency and a relatively recent version is required to trim both ends.
```{r}
if (grep("Windows", osVersion, ignore.case = TRUE)) {
  cutadapt <- "wsl cutadapt"
} else {cutadapt <- "cutadapt"}

system(paste(cutadapt,"--version"))

```


Trim primers with cutadapt
```{r message=FALSE, warning=FALSE, cache=TRUE}

cores <- as.numeric(future::availableCores())


path.cut <- file.path(path, "cutadapt")
if(!dir.exists(path.cut)) dir.create(path.cut)
fnFs.cut <- file.path(path.cut, basename(fnFs))
fnRs.cut <- file.path(path.cut, basename(fnRs))


FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC) 
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC)  

tictoc::tic()

if ((grep("Windows", osVersion, ignore.case = TRUE))) { #Run on windows subsystem for linux if OS is windows
  for(i in seq_along(fnFs)){
  system2("wsl", args = c("cutadapt",R1.flags, R2.flags, "--minimum-length", 50,
                             "-n", 2, # -n 2 required to remove FWD and REV from reads
                             paste("--cores=",cores,sep = ""),
                             "-o", str_replace(fnFs.cut[i],"D:/","/mnt/d/"), "-p", str_replace(fnRs.cut[i],"D:/","/mnt/d/"), # output files, path fixed for
                             str_replace(fnFs.filtN[i],"D:/","/mnt/d/"), str_replace(fnRs.filtN[i],"D:/","/mnt/d/")), # input files, path fixed for wsl
          stdout = as.character(file.path(path.cut,"cutadapt.log"))) #save log
  } 
} else{ #Run normally if OS is not Windows
  for(i in seq_along(fnFs)){
  system2("cutadapt", args = c(R1.flags, R2.flags, "--minimum-length", 50,
                             "-n", 2, # -n 2 required to remove FWD and REV from reads
                             paste("--cores=",cores,sep = ""),
                             "-o", fnFs.cut[i], "-p", fnRs.cut[i], # output files
                             fnFs.filtN[i], fnRs.filtN[i]),stdout = as.character(file.path(path.cut,"cutadapt.log"))) # input files
  } 
}

tictoc::toc()
```



```{r, cache=TRUE}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.cut[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.cut[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.cut[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.cut[[1]]))
```

List the files processed by cutadapt
```{r, cache=TRUE}
cutFs <- sort(list.files(path.cut, pattern = "L001_R1_001.fastq.gz", full.names = TRUE))
cutRs <- sort(list.files(path.cut, pattern = "L001_R2_001.fastq.gz", full.names = TRUE))
```

Get sample names
```{r, cache=TRUE}
get.sample.name <- function(fname) strsplit(basename(fname), "_\\d*_S")[[1]][1]
sample.names.f <- unname(sapply(cutFs, get.sample.name))
sample.names.r <- unname(sapply(cutRs, get.sample.name))
head(sample.names.f)
head(sample.names.r)
```

## Examine read quality

Plot examples of forward read quality
```{r fig.height=4, fig.width=8, out.width="100%"}
plotQualityProfile(cutFs[1:2])
```

Plot examples of reverse read quality
```{r fig.height=4, fig.width=8, out.width="100%"}
plotQualityProfile(cutRs[1:2])
```



## Write a QIIME2 manifest
```{r}

manifest.f <- data.frame(sample.id = sample.names.f,
                       absolute.filepath = cutFs,
                       direction = "forward")

manifest.r <- data.frame(sample.id = sample.names.r,
                       absolute.filepath = cutRs,
                       direction = "reverse")

manifest <- bind_rows(manifest.f,manifest.r) %>% 
  rename("sample-id" = sample.id, "absolute-filepath" = absolute.filepath)

if ((grep("Windows", osVersion, ignore.case = TRUE))) {
 manifest <- manifest %>% 
   mutate(`absolute-filepath` = str_replace(`absolute-filepath`,"D:/","/mnt/d/"))
}

write_csv(manifest, file.path(here("data/manifest.txt")))
```

